### PR TITLE
Making this work better.

### PR DESCRIPTION
**Pull request by atrull**
Several major changes here, and I am trying to sum them up accurately. If you are an existing user of this terraform module PLEASE READ CAREFULLY.

Switching to a regular userdata format which lets us avoid the tabbing nonsense of yaml, which broke the wireguard config each time(only the last Peer was valid!).

Moved EIP outside of the module to allow the module to be used for additional ELB mode (private ip) server. (breaking change #1 - please see the EIP example)

Launch configuration can now drive Autoscaling group creation allowing for terraform to add new vpn users to the config by switching out the scaling group if sized correctly.

Outputs are changed to remove the previous in-module EIP usage but also give the two security groups the module makes, and the autoscaling group name.

Ubuntu ami is dynamic via the ami search, working for any region with "ubuntu/images/hvm-ssd/ubuntu-*-16.04-amd64-server-*" AMIs

Fully Terraform 12 now. Readme cleanup and extended examples to cover EIP and ELB setups.

Variables Added:

Switching to a variable for the wireguard server interface which, which should be a user choice especially as it regards the client ips.. it takes an IP + Subnet Mask (x.x.x.x/N). The clients MUST be within the range of the mask. Several other new new wg_ and asg_ variables to tune the setup including the server port via wg_server_port.

eip_id variable added so we can tie it back correctly for EIP style usage. (breaking change #1 - please see the EIP example)

additional_security_group_ids added to allow more flexibility of SG usage and especially to allow health checks on port 22 (ssh).

associate_public_ip_address added so we can specify whether or not to take public IP (needed for EIP mode with public subnet)

instance_type variables added so we can specify larger or smaller machine types.

Variables Renamed:

Renamed public_subnet_ids to subnet_ids to more accurately address the usage in the case of ELB mode.

### CODE REVIEW

**Comment by atrull** [comment]
@jmhale - I'd like to propose moving the EIP outside of the module and supplying its id like the subnets and ranges. This will make it cleaner whether this is run being an EIP on an instance or behind an ELB..

**Comment by atrull** [comment]
> @jmhale - I'd like to propose moving the EIP outside of the module and supplying its id like the subnets and ranges. This will make it cleaner whether this is run being an EIP on an instance or behind an ELB..

I've included my changes for how to make this work. Also extended the examples in README.

**Comment by atrull** [comment]
😭 is my PR such low quality, or is the author on holiday ?

File: wireguard-securitygroups.tf [hunk 1]
```diff
@@ -3,20 +3,27 @@ resource "aws_security_group" "sg_wireguard_external" {
...
-    from_port   = 51820
-    to_port     = 51820
+    from_port   = "${var.wg_server_port}"
+    to_port     = "${var.wg_server_port}"
     protocol    = "udp"
     cidr_blocks = ["0.0.0.0/0"]
   }
 
+  ingress {
+    from_port   = 22
```
Comment (jmhale) [inline wireguard-securitygroups.tf]: We shouldn't be exposing port 22 to the world. Need to remove this.

File: variables.tf [hunk 2]
```diff
@@ -21,6 +16,32 @@ variable "wg_client_public_keys" {
...
+}
+
+variable "wg_server_port" {
+  default = 51820
+  description = "Port for the vpn server"
+}
+
+variable "eip_id" {
+  default = "set_me_or_lose_me"
+  description = "ID of the EIP allocation if the defailt remain the eip instructions in userdata are ignored"
```
Comment (jmhale) [inline variables.tf]: This description could use some clean up. Hard to tell what the var is used for on first read through.

File: wireguard-securitygroups.tf (hunk 1 above)
Comment (atrull) [reply wireguard-securitygroups.tf]: Something I left in for debuggery - will get it sorted.

File: variables.tf (hunk 2 above)
Comment (atrull) [reply variables.tf]: Yeah I really should rewrite that.

Comment (atrull) [reply variables.tf]: I've replaced 'set_me_or_lose_me' with 'disabled' which is far more unambiguous and improved the description.

**Comment by atrull** [comment]
@jmhale changes ready for your review :-)

**Comment by atrull** [comment]
> THANK YOU DEAR GOD FOR DOING THIS

Thanks, nobody ever said that about my PRs before.

**Comment by atrull** [comment]
I'm trying to get this to actually WORK in tf 0.12 format, that'll take a bit of time.

**Comment by atrull** [comment]
ready for more testing with tf 0.12. current state is that this works on my implementation..

**Comment by atrull** [comment]
@danehrlich1 I'd appreciate your testing too, FWIW!

**Comment by danehrlich1** [comment]
@atrull I can get it all to work. Terraform has a command from the CLI that will do  most of the work for you. Any other errors you run into...error message should be pretty obvious (at least in 12 much better than 11!)

**Comment by atrull** [comment]
I'm back from some vacation @jmhale - I've made this a bit more mature, in the last few days. It's working good in my test case (windows and osx clients). I'm actually needing to put it into production shortly to negate some infosec issues we have with 'open bastions' :D 🔢 

Would like another review and, maybe is it good enough to merge ?

**Comment by jmhale** [comment]
@atrull I think we're getting closer. I had to resolve some conflicts that I introduced, but I was able to test out the EIP config again and everything is still working as expected.

For the NLB config, I had a question:

- How are you testing using the NLB? In my testing, the NLB wouldn't mark the target at healthy, because it doesn't support UDP port checks, and the only port that the SG is allowing is 51820/UDP. 

I had to add a SG rule for, and override the health check to point to, 22/TCP, in order to get the LB to mark the target as healthy and start forwarding WG traffic. 

It looks like that's the only TCP port listening on the instance, so we may have to do that no matter what, but we'd need to add a rule to the module for the SG to allow 22/TCP from the VPC CIDR. Not the greatest solution, so I'd be interested in alternatives.


Some improvements for using the NLB scenerio:

- If using a NLB, a good principle of least privilege uplift would be to not include the creation of the IAM policy, role, and attachments, since those are only used to associate an EIP, which isn't being used in this scenario.

- I'm thinking that we should add the `persistent-keepalive` option to the WG config, if the NLB option is used. Since in this config, the WG instance is sitting on private subnets, and would need to egress a NAT (instance|gateway), we don't want the NAT to eat the connection if there isn't a lot of traffic. I haven't had a whole lot of time to test and see if this is actually an issue, but just guessing that it might be. https://www.wireguard.com/quickstart/#nat-and-firewall-traversal-persistence

**Comment by atrull** [comment]
@jmhale I've made a few updates for ELB scenarios, heading to bed now but will go through your response early next week. Thanks for looking at this! And yes, the port 22 health check thing, heh! :-)

**Comment by atrull** [comment]
> @atrull I think we're getting closer. I had to resolve some conflicts that I introduced, but I was able to test out the EIP config again and everything is still working as expected.
> 
> For the NLB config, I had a question:
> 
> * How are you testing using the NLB? In my testing, the NLB wouldn't mark the target at healthy, because it doesn't support UDP port checks, and the only port that the SG is allowing is `51820/UDP`.
> 
> I had to add a SG rule for, and override the health check to point to, `22/TCP`, in order to get the LB to mark the target as healthy and start forwarding WG traffic.
> 
> It looks like that's the only TCP port listening on the instance, so we may have to do that no matter what, but we'd need to add a rule to the module for the SG to allow `22/TCP` from the VPC CIDR. Not the greatest solution, so I'd be interested in alternatives.

Two thoughts, we either document the need to do this, or just stick it in the module, or ??. For the sake of being against ignorance, I think documenting it is better.. for user's to understand integration, I've put something in the README, see how you feel about it ?

> Some improvements for using the NLB scenerio:
> 
> * If using a NLB, a good principle of least privilege uplift would be to not include the creation of the IAM policy, role, and attachments, since those are only used to associate an EIP, which isn't being used in this scenario.

Agreed, I need to look at a clean way to make terraform apply less for once case but more for the other. I'd like your (or anyone else watchin's) thoughts on how!

> * I'm thinking that we should add the `persistent-keepalive` option to the WG config, if the NLB option is used. Since in this config, the WG instance is sitting on private subnets, and would need to egress a NAT (instance|gateway), we don't want the NAT to eat the connection if there isn't a lot of traffic. I haven't had a whole lot of time to test and see if this is actually an issue, but just guessing that it might be. https://www.wireguard.com/quickstart/#nat-and-firewall-traversal-persistence

I happen to think this is a good idea for any client config (EIP or ELB setups) for this module, because from the perspective of the hosts (servers) for both the NAT / ELB and EIP scenarios, is that that IP that is public isn't directly assigned to the server, not actually on a local eth0 interface, it's all AWS networking black box done by the network gateway.

**Comment by jmhale** [comment]
- For documenting allowing `22/TCP` rather than automatically including it in the code, I agree. I think that the NLB config is a bit more of an advanced use case, so I'm willing to have it be less "batteries included" for the sake of it being more secure. What you've included in `README` is good IMO.

- For `persistent-keepalive`, yeah that makes sense. I don't see any harm in including it in either config scenerio. WG docs suggest 25 seconds, which we can start with that, and tune if it's an issue.

- Conditionally applying resources is a lot cleaner in TF v0.12. 

You can conditionally set the `iam_instance_profile` param in the launch config like this:
```
  iam_instance_profile        = (var.eip_id != "disabled" ? aws_iam_instance_profile.wireguard_profile[0].name : null)
```

And then conditionally create each of the IAM resources, by including the `count` param:
```
count       = (var.eip_id != "disabled" ? 1 : 0)
```

Although, once you start using `count`, you have to reference those resources by index. Since it's either one or zero resources, you can safely use `[0]` everywhere.

**Comment by atrull** [comment]
@jmhale I've committed these improvements and tested them on my use case POC (all good). Can you try them on yours ?

**Comment by atrull** [comment]
@jmhale I've updated the opening comment of this PR to reflect most if not all the relevant changes, which may help you for changelog bits.

**Comment by jmhale** [comment]
This looks good. I'm going to merge it to a develop branch, so I can do the CHANGELOG work, before it lands in master. Thanks for your work on this!

**Pull request merged**